src/EditorFeatures/VisualBasic/Highlighting/KeywordHighlightingHelpers.vb

```diff
@@ -8,127 +8,112 @@ Imports Microsoft.CodeAnalysis.VisualBasic.Syntax
-                        Dim text = node.SyntaxTree.GetText(cancellationToken)
-                        Dim previousLine = text.Lines.IndexOf(previousToken.SpanStart)
-                        Dim awaitLine = text.Lines.IndexOf(.AwaitKeyword.SpanStart)
-
-                        If previousLine = awaitLine Then
-                            Dim index = highlights.FindIndex(Function(s) s.Contains(previousToken.Span))
-                            If index >= 0 Then
-                                Dim span = highlights(index)
-                                highlights(index) = TextSpan.FromBounds(span.Start, .AwaitKeyword.Span.End)
-                                handled = True
-                            End If
-                        End If
-                    End If
+        Friend Iterator Function HighlightRelatedAwaits(thisnode As SyntaxNode, cancellationToken As CancellationToken) As IEnumerable(Of TextSpan)
+            If cancellationToken.IsCancellationRequested Then Return
+
+            Dim nodes As New LinkedList(Of SyntaxNode)
+            nodes.AddFirst(thisnode)
+            While nodes.Count > 0
 
-                    If Not handled Then
-                        highlights.Add(.AwaitKeyword.Span)
-                    End If
-                End With
-            End If
+                Dim node = nodes(0) : nodes.RemoveFirst()
 
-            For Each child In node.ChildNodes()
-                If Not TypeOf child Is LambdaExpressionSyntax Then
-                    HighlightRelatedAwaits(child, highlights, cancellationToken)
+                If cancellationToken.IsCancellationRequested Then
+                    Continue While
                 End If
-            Next
-        End Sub
-
-        Private Sub HighlightRelatedYieldStatements(Of T)(node As SyntaxNode, highlights As List(Of TextSpan))
-            If TypeOf node Is YieldStatementSyntax Then
-                With DirectCast(node, YieldStatementSyntax)
-                    highlights.Add(.YieldKeyword.Span)
-                End With
-            Else
-                For Each child In node.ChildNodes()
-                    If Not TypeOf child Is LambdaExpressionSyntax Then
-                        HighlightRelatedYieldStatements(Of T)(child, highlights)
-                    End If
-                Next
-            End If
-        End Sub
+
+                If TypeOf node Is AwaitExpressionSyntax Then
+                    With DirectCast(node, AwaitExpressionSyntax)
+                        Yield .AwaitKeyword.Span
+                    End With
+                End If
+
+                Dim children = node.ChildNodes.Where(Function(child) TypeOf child IsNot LambdaExpressionSyntax)
+                nodes.AddRangeAtHead(children)
+
+            End While
+        End Function
+
+        Private Iterator Function HighlightRelatedYieldStatements(Of T)(thisnode As SyntaxNode) As IEnumerable(Of TextSpan)
+            Dim nodes As New LinkedList(Of SyntaxNode)
```

Comment:
@jasonmalinowski  I'm using a `LinkedList` to help preserve the order of evaluation, due the pre-pull request recursion.
Don't know if there is a better method.